autogluon/utils/tabular/features/generators/abstract.py

```diff
@@ -0,0 +1,587 @@
+# TODO: Add log of # of observation counts to high cardinality categorical features
+class AbstractFeatureGenerator:
+    """
+    Abstract feature generator implementation from which all AutoGluon feature generators inherit.
+    The purpose of a feature generator is to transform data from one form to another in a stateful manner.
+    First, the generator is initialized with various arguments that dictate the way features are generated.
+    Then, the generator is fit through either the `.fit()` or `.fit_transform()` methods using training data typically in pandas DataFrame format.
+    Finally, the generator can transform new data with the same initial format as the training data through the `.transform()` method.
+
+    Parameters
+    ----------
+    features_in : list, default None
+        List of feature names the generator will expect and use in the fit and transform methods.
+        Any feature in an incoming DataFrame that is not present in features_in is dropped and will not influence the transformation logic.
+        If None, infer during fit from the _infer_features_in method.
+        Equivalent to feature_metadata_in.get_features() post-fit.
+    feature_metadata_in : FeatureMetadata, default None
+        FeatureMetadata object corresponding to the training data input features.
+        If None, infer during fit from the _infer_feature_metadata_in method.
+        Any features not present in features_in (if provided) will be removed from feature_metadata_in.
+    post_generators : list of FeatureGenerators, default None
+        FeatureGenerators which will fit and transform sequentially after this object's transformation logic, feeding their output into the next generator's input.
+        The output of the final FeatureGenerator will be the used as the transformed output.
+    pre_enforce_types : bool, default False
+        If True, the exact raw types (int64, float32, etc.) of the training data will be enforced on future data, either converting the types to the training types or raising an exception if unable.
+        This is important to set to True on the outer feature generator in a feature generation pipeline to ensure incorrect dtypes are not passed downstream, but is often redundant when used on inner feature generators inside a pipeline.
+    pre_drop_useless : bool, default False
+        If True, features_in will be pruned at fit time of features containing only a single unique value across all rows.
+    post_drop_duplicates : bool, default False
+        If True, a DropDuplicatesFeatureGenerator will be appended to post_generators.
+        This feature generator will drop any duplicate features found in the data, keeping only one feature within any duplicate feature sets.
+        Warning: For large datasets with many features, this may be very computationally expensive or even computationally infeasible.
+    reset_index : bool, default False
```

Comment:
Is there a good reason for autogluon to try and preserve pandas indices?  I feel like these are a bit of pain to preserve, and mainly used for joining data from multiple sources which seems not relevant to ML.  It could be useful to preserve index for test-predictions, but I'm not sure it's necessary as long as the order of predictions matches order of test datapoints.

Reply:
It is indeed massively painful. However, it is dangerous to ignore the index due to a variety of edge case scenarios. This can be a separate discussion in future if we want to have a smarter way of handling indices, but the feature generators handle them about as well as can be hoped for I think.